Another attempt of support pty on windows with ConPTY api

photostorm opened the pull request:
This is an another attempt to support pseudo-terminal on windows using ConPty API (should resolve issue https://github.com/creack/pty/issues/95)

This pull request does introduce major api change for Start method (*os.File -> Pty),

I have tested this code with remote terminal application that I wrote. It would be great to get some additional testing with other use cases.

photostorm committed:
new windows support

photostorm committed:
Update README.md

photostorm committed:
temporary module path

photostorm committed:
Revert "temporary module path"

This reverts commit 2aa609e6e2086afec572193405d26ab1033e960f.

photostorm committed:
made exec cmd struct private

photostorm committed:
Revert "Update README.md"

This reverts commit d44d7f77e385f0bd1557aed33da25b93a491c918.

pete-woods commented:
This looks very promising, I will try it out to see if it works in our system

pete-woods commented:
I just tried this out, and works perfectly for my use case (an SSH server that must work on UNIX and Windows) 👍

mjudeikis commented:
Trying this now. Would be really nice if this works on go 1.18 too :) considering repo is still pointing to 1.13 :D

pete-woods commented:
It worked on Go 1.18 for me

mjudeikis commented:
It just passed our ci too. Worked like a charm too :)

On Thu, Jul 28, 2022, 7:49 PM Pete Woods ***@***.***> wrote:

> It worked on Go 1.18 for me
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/creack/pty/pull/155#issuecomment-1198395129>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYTREGNBTBOT5HCXFTO3ATVWK23LANCNFSM54T43GUQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

pete-woods commented:
@creak if you have time to look at this, it would be most appreciated, I found the API break to be extremely unintrusive (in that I didn't need to change any code)

creack commented:
I’ll take a look over the weekend. Quite a few tickets in backlog, sorry
about that.

On Thu, Jul 28, 2022 at 10:51 AM Pete Woods ***@***.***>
wrote:

> @CrEaK <https://github.com/CrEaK> if you have time to look at this, it
> would be most appreciated, I found the API break to be extremely
> unintrusive (in that I didn't need to change any code)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/creack/pty/pull/155#issuecomment-1198396721>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7LZMFPY36MECEUL42SDDVWK3CBANCNFSM54T43GUQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
-- 
Guillaume J. Charmes

pete-woods commented:
![minions-please](https://user-images.githubusercontent.com/152569/184148180-07b3b704-4902-4728-ad40-beb1c75313c4.jpeg)

mjudeikis commented:
![image](https://user-images.githubusercontent.com/3225744/184149391-59459256-f649-4869-80bc-840b683c66c2.png)
Please help to remove wild forks!

gabemarshall commented:
Worked for me on 1.18 and 1.17.1, thanks @photostorm 👏👏👏

pete-woods commented:
@creack Is there anything we can do to help with this PR getting reviewed? Windows support is a really big deal for a bunch of users of your library.

mohammed90 commented:
I'm happy this PR exists. However, it introduces a breaking change: `*os.File` -> `Pty`. The func `Start` and `StartWithAttrs` return `*os.File` currently, but the PR changes this to type `Pty`.

photostorm commented:
There should be high bar to justify incompatible changes to the interface. Due to the return type of Start, I do not see a way to pass the handles. I think this API change would make it possible for other systems in future to be added.

photostorm commented:
ok, I will look into it. Thanks for your feedback.

photostorm commented:
@kr I do not think I can used that. there is window handle that I need access to also.

photostorm committed:
cleanup some unused fields

pete-woods commented:
That seems like a plausible strategy to me, but without input from @creack I don't thing we can realistically progress

sabattle commented:
@creack have you gotten the chance to review this PR? My team could definitely use this. At the moment we are relying on @photostorm's fork (which has worked perfectly so far)

tflpd commented:
@creack It would be deeply appreciated if this PR could be merged 🙏🏼

Naatan committed:
Don't close a pty that was never started as it causes a panic

photostorm committed:
Merge pull request #1 from ActiveState/master

Don't close a pty that was never started as it causes a panic

carghai commented:
The code looks good! really hoping this gets merged. This will help my project so much!

CharlesChiuGit commented:
Seems like the author has abandoned the project, maybe it's better to create another fork?

creack commented:
I am still around, but indeed real life has been on the way lately. I have
quite a lot of pry pr/issues pinned, I’ll address them as soon as I can.
This kind of thing require quite a lot of time.

On Wed, Jun 7, 2023 at 10:50 PM Charles Chiu ***@***.***>
wrote:

> Seems like the author has abandoned the project, maybe it's better to
> create another fork?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/creack/pty/pull/155#issuecomment-1581808414>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD7LZOSHUEFIYWVM2JYF2LXKE4Y3ANCNFSM54T43GUQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
-- 
Guillaume J. Charmes

CharlesChiuGit commented:
tks for the reply!

a-urth commented:
looking forward to get this merged!

Naatan commented:
@photostorm since this PR has yet to be merged (eagerly awaiting the day!) and since your repo has issues turned off the best place I can think to report an issue is on this thread.

I've been trying to figure out why the windows implementation has a tendency to produce superfluous output. It seems to often start with the path of the executable, followed by an ALERT signal (unicode U+0007). It seems to do this almost every single invocation, but depending on the performance of the host as well as the process being ran it may be cut off differently, and may even appear intermixed with regular output.

I've created a repro of the issue here:

https://github.com/ActiveState/termtest/blob/a695852f3f6479d5fbed93abd19562525c4045cf/test/pty/pty_test.go

Note this repro itself has a bug which I haven't bothered to further debug at this time as it is a red herring. But basically it will never exit, as the last read just hangs. I'm not getting this with our regular use of the lib so won't bother digging on that.

Anyway, if you run this test you will get something like this (again: host and process performance may affect this):

```
Read: 








]0;C:\Program Files\Go\bin\go.exeRead: hello
```

The first read seems entirely superfluous, though note it's not guaranteed to happen in the first read, or even at all.

It seems there is some sort of race condition happening somewhere, but so far I've been unable to nail it down.

Happy to report this as an issue on your repo if you'd be willing to open it up to issues, as I don't mean to distract from this PR. Nor do I think this PR should be blocked by this bug; bugs are gonna happen one way or another, and this one isn't a hard blocker in my opinion.

photostorm commented:
@Naatan They should be open now.

Naatan commented:
@photostorm thanks! Opened a ticket on your repo instead 🙏

photostorm committed:
cleanup some used code, added code to monitor the spawned process

asjdf commented:
Over a year since this repository was last active, it's time to maintain a more actively updated fork.

pete-woods commented:
Much enthusiasm for a v2 with Windows support, and a swift adoption by the community.

Given projects like Docker (with huge numbers of dependents) depend on it, it should make the transition easier.

An x/windows package had a break a few years ago and it was a total PITA for anyone depending on Docker.

Naatan commented:
I'd be perfectly happy with a v2 branch. I could not care less which version is determined "stable"; it's about having the assurance that there is an intent for the code to be maintained, as well as a central place to follow further development. 

Thankfully @photostorm has been extremely responsive, so those concerns are partly addressed already, but ultimately that makes the current situation such that we basically have an unofficial fork already.

pidgeon777 commented:
Hopefully this branch will be merged to solve this issue:

https://github.com/jesseduffield/lazygit/issues/1453

rvhelden commented:
Really waiting for this to merge so lazygit pagers will be usable for windows

https://github.com/jesseduffield/lazygit/issues/1453

RollsChris commented:
I'm also now looking for this to work. thanks :)

creack committed:
Merge branch 'master' into master

creack commented:
I am not familiar with Windows nor ConPTY, but looking at the code in more detail, I see it relies on globals to manage "a" pty. Doesn't this mean the Windows implementation is limited to a single child process? That would be quite a gap from the Unix version.

photostorm committed:
fixed type issue from recent merge

photostorm committed:
fixed unit tests

creack committed:
Minor cosmetic fixes, add few comments/questions.

pete-woods commented:
I'm hoping that's an oversight. I was expecting to see some kind of global map, to work around having an API break

photostorm commented:
The only thing that I'm pretty sure I have there that is global is the reference to the system DLL. Is there a particular area you were referring to?

photostorm commented:
@pete-woods Yeah I know I tried looking into using the UNIX socket support to prevent API breaks that @kr mentioned earlier. It was just I could not think of a good way of keeping up with the handlers quite yet. So if anyone has any ideas, that would be helpful.

mohammed90 commented:
The API breakage was an issue with v1. However, the latest commits changed the module path to include `/v2`, which allows for breaking change per both semver and Go module rules.

photostorm committed:
moved deadline method to FdHolder

photostorm committed:
added DeadlineHolder

pete-woods commented:
Great, if we're going the v2 route 🎉

creack commented:
@photostorm I can't get Setsize to work with this PR, when using it, it sets cols/rows to 1, regardless of the value I attempt to set.

This results in InheritSize not working either and any children spawn up to have either size 0/0 or 1/1.

Any idea?

Note that before calling Setsize, Getsize returns the expected value.

Sample code:

```go
package main

import (
        "fmt"
        "log"
        "os"
        "os/exec"
        "strconv"
        "time"

        pty "github.com/creack/pty/v2"
        "golang.org/x/term"
)

func child(n int) {
        f, err := os.OpenFile("child.logs", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
        if err != nil {
                log.Fatalf("[%d] Errir openning logs file: %s.", n, err)
        }
        log.SetOutput(f)
loop:
        isTTY := term.IsTerminal(1)
        w, h, err := pty.Getsize(os.Stdout)
        if err != nil {
                log.Fatalf("[%d] Error getting size: %s.", n, err)
        }
        log.Printf("[%d] isTTY: %t, width: %d, height: %d\n", n, isTTY, w, h)
        time.Sleep(1 * time.Second)
        goto loop
}

func startChild(n int) pty.Pty {
        c := exec.Command(os.Args[0], strconv.Itoa(n))

        ptmx, err := pty.Start(c)
        if err != nil {
                log.Fatalf("[%d] Error starting child in pty: %s.", n, err)
        }

        return ptmx
}

func assertChildSize(n int, child pty.Pty, expected *pty.Winsize) {
        childSize, err := pty.GetsizeFull(child)
        if err != nil {
                log.Fatalf("[%d] Error getting the child size: %s.", n, err)
        }
        if childSize.Cols != expected.Cols || childSize.Rows != expected.Rows {
                log.Fatalf("[%d] Unexpected child size. %v != %v.", n, childSize, expected)
        }
}

func test() error {
        // If we have an argument, we are in a child process.
        if len(os.Args) > 1 {
                // Parse the argument as the child number.
                n, err := strconv.Atoi(os.Args[1])
                if err != nil {
                        log.Fatalf("Error parsing child number: %s.", err)
                }
                child(n)
                return nil
        }

        // If we dont have an argument, we are in the parent. Spin up a child and attempt to resize it.
        initialSize, err := pty.GetsizeFull(os.Stdout)
        if err != nil {
                log.Fatalf("Error getting the initial size: %s.", err)
        }
        fmt.Printf("Initialsize: %v\n", initialSize)

        child := startChild(0)

        // if err := pty.InheritSize(os.Stdout, child); err != nil {
        //      log.Fatalf("Error inheritsize: %s.", err)
        // }
        // assertChildSize(0, child, initialSize)

        if err := pty.Setsize(child, &pty.Winsize{Rows: 20, Cols: 20}); err != nil {
                log.Fatalf("Error setsize: %s.", err)
        }
        assertChildSize(0, child, &pty.Winsize{Rows: 20, Cols: 20})

        log.Printf("Success.")

        return nil
}

func main() {
        if err := test(); err != nil {
                log.Fatal(err)
        }
}
```

```sh
go get github.com/creack/pty/v2@photostorm-master-creack
GOOS=windows go build -o /tmp/foo.exe
```

creack committed:
Merge branch 'master' into master

photostorm commented:
I will take a look when I get some time later this week.

creack commented:
Minor-ish issue: closing the pty twice causes a heap violation resulting in the parent process to die (exit status 0xc0000374).

The case happens often when dealing with network where multiple thing can cause the server to want to kill the process / close the pty.

creack commented:
Another issue is that I can't get a shell to properly work :(, it seems to be working at first, but as soon as I reach ~50% of the terminal, the cursor starts to jump around randomly.

https://asciinema.org/a/uNCRfSYmuuwYqoqu9Q4a7SEOn

A regular cmd or pwsh without the lib wrapping it works as expected.

creack commented:
confirmed, when manually setting the size, it works fine, with one caveat, still having an issue with `MaxPhysicalWindowSize` though, I can't figure out how to change this value.

Some programs (like vim) use that as upper bound, which results in the wrong size.

creack commented:
As OpenSSH suffers from the same `MaxPhysicalWindowSize` issue, once we fix the crash on `.Close()` and when we have a reliable way to set size, we can merge.

photostorm commented:
I have a fix for the close issue. I am currently investigating sizing behavior. It is interesting that the setSize function seems to report to success, but when using the GetConsoleScreenBufferInfo method, it reports invalid handle error.

pete-woods commented:
> I have a fix for the close issue. I am currently investigating sizing behavior. It is interesting that the setSize function seems to report a success, but when using the GetConsoleScreenBufferInfo method, it reports an invalid handle error.

Thankyou for continuing to take time on this issue

adrivn commented:
Any updates? Thank you for dedicating your time to this issue

photostorm commented:
> Any updates? Thank you for dedicating your time to this issue

Right now, I'm still working on fixing the invalid handle issue with get size. It seemed I was using the wrong type of handle for the console screen buffer info. It should be the stdout handle. However, that also seems to be reporting an invalid handle. so I just been researching to try to figure out the correct way of handling the situation. Any help would be appreciation.

theclapp committed:
Don't close consoleR & consoleW in Windows

pty_windows.go closed consoleW and consoleR in the initial open(), with
the comment "These pipes can be closed here without any worry."

This isn't actually true.

If you interact with the terminal via Go code (i.e. don't immediately
fork a new process, but use (for example) github.com/mvdan/sh), you need
those pipes to stay open, and to close them yourself.

theclapp committed:
Make tests pass

photostorm committed:
Merge pull request #7 from HuckRidgeSW/master

Don't close consoleR & consoleW in Windows

pete-woods commented:
Thankyou for continuing to plug away at this

iyzyi commented:
I don't mean to disrupt the discussion about this PR, but perhaps you could take a look at [aiopty/pty/conpty/conpty_windows.go](https://github.com/iyzyi/aiopty/blob/main/pty/conpty/conpty_windows.go) and [aiopty/term/term_windows.go](https://github.com/iyzyi/aiopty/blob/main/term/term_windows.go).

kcmvp commented:
thank you all very much for implementation tty on windows!
any update on this?

pidgeon777 commented:
Great news! 👍

creack commented:
Getting an error when running the tests (go1.22.3)
```
.\io_test.go:33:32: cannot use int(ptmx.Fd()) (value of type int) as syscall.Handle value in argument to syscall.SetNonblock
.\io_test.go:69:32: cannot use int(ptmx.Fd()) (value of type int) as syscall.Handle value in argument to syscall.SetNonblock 
```

creack commented:
When commenting out the SetDeadline section which cause the build to fail, I get test errors:

```
--- FAIL: TestSetsize (0.01s)e.
    doc_test.go:102: Unexpected Getsize X result after Setsize: 1 != 0.
    doc_test.go:103: Unexpected Getsize Y result after Setsize: 1 != 0.
    doc_test.go:104: Unexpected Getsize Rows result after Setsize: 2 != 1.
    doc_test.go:105: Unexpected Getsize Cols result after Setsize: 2 != 1.
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
--- FAIL: TestOpen (0.01s)
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
--- FAIL: TestGetsizeFull (0.02s)
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
--- FAIL: TestOpenByName (0.02s)
    doc_test.go:38: Failed to open tty file: open |0: The filename, directory name, or volume label syntax is incorrect..
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
--- FAIL: TestName (0.02s)
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
--- FAIL: TestGetsize (0.02s)
    helpers_test.go:24: Unexpected error from pty Close: Access is denied..
exit status 0xc0000374
FAIL    github.com/creack/pty/v2        0.357s
```

photostorm commented:
I still working on fixing invalid handle with get size, but just been busy with work.

kareemmahlees commented:
Any update on this ?

GitMurf commented:
> Really waiting for this to merge so lazygit pagers will be usable for windows
> 
> [jesseduffield/lazygit#1453](https://github.com/jesseduffield/lazygit/issues/1453)

This is why I am here too 😢 Any updates? It seems without this, LazyGit cannot allow using a custom pager like delta for git diffing things. Would be super useful. Thanks so much for everyone's work on this!

pidgeon777 commented:
Hope is still high for this PR

photostorm commented:
I am still here. Just been busy with my work. I hope to get back to soon. If any one can help me with looking into invalid handle problem that would be great.

wellcomez commented:
> I am still here. Just been busy with my work. I hope to get back to soon. If any one can help me with looking into invalid handle problem that would be great.

it works fine for , thanks.  @photostorm 
@creack  it looks ok, why not?

creack commented:
Running the tests yield somme failures, cf my last message. @photostorm mentioned he still needs some work on this.
When I find some time myself, I’ll try to wrap it up.

wellcomez commented:
> Running the tests yield somme failures, cf my last message. @photostorm mentioned he still needs some work on this. When I find some time myself, I’ll try to wrap it up.

compared with *nix, I find main issue is its performace is too poor under window
@photostorm  @creack

pidgeon777 commented:
> > Running the tests yield somme failures, cf my last message. @photostorm mentioned he still needs some work on this. When I find some time myself, I’ll try to wrap it up.
> 
> compared with *nix, I find main issue is its performace is too poor under window @photostorm @creack

Could you please add more details?

wellcomez commented:
> > > Running the tests yield somme failures, cf my last message. @photostorm mentioned he still needs some work on this. When I find some time myself, I’ll try to wrap it up.
> > 
> > 
> > compared with *nix, I find main issue is its performace is too poor under window @photostorm @creack
> 
> Could you please add more details?

https://github.com/wellcomez/lspvi/releases/tag/v100test my project
1. ./lspvi --gui open browser  open url in browser  https://localhost:13000 or 1300x 
2. ./lspvi in windows terminal (not cmd.exe) 
web is slower than terminal 
3. but under linux/macos, web mode is as fast as terminal
@pidgeon777

lukasbergman commented:
Any updates on this? @creack @photostorm

XhstormR commented:
ping

vaspadi commented:
ping
